README
TAP utilities (John Elliott, 27 March 2011)
===========================================

  General note: All these programs except specform, unspecform, mkp3fs and
sna2dsk operate on files in the .TAP format. They also support .ZXT, which 
is a .TAP file with a +3DOS header, used by TAPROM on the +3.

  When compiling this software, it helps to have LibDsk 
<http://www.seasip.demon.co.uk/Unix/LibDsk/> installed. Otherwise those
programs which output disk images (tap2dsk, sna2dsk and mkp3fs) will be 
severely restricted in their output capabilities, and dsk2tap won't work at 
all.

  Programs in this archive are:

specform
--------
  specform adds a +3DOS header to one or more files, or corrects the checksum 
in an existing +3DOS header.

  Syntax is: specform { -F } { -a addr } <file> <file> ...

  If the -F option is not present, a +3DOS header will be added to each file
(even those that have one already) and the result saved as a file with the 
same name, but ending in .zxb.
  If the -a option is present, the load address specified will be used in
the +3DOS header added to each file. Otherwise, a load address of 0 will be
used.
  If the -F option is present, each file must be a +3DOS file. The header
checksum will be recalculated and saved back to the same file.

unspecform   (unspecfm in DOS)
----------

  unspecform removes the +3DOS header from a single file. It also removes 
any bytes beyond the end of the file (usually caused by the process of copying
+3DOS files from CP/M media).

  Syntax is: unspecform { input { output }}

  If the input and output filenames aren't specified, standard input and 
output are used. 


bin2bas
-------

  bin2bas converts wraps a binary file (with or without a +3DOS header) in
a BASIC program consisting of two lines. The first line is a REM statement
containing the binary file, and the second is a BASIC USR command to execute
the code in the REM:

	LET x = USR(5 + PEEK 23635 + 256 * PEEK 23636)

  Your binary file should therefore begin with Z80 code. This should be 
position-independent, and follow the usual Spectrum BASIC calling conventions:

* At entry, BC holds the load address.
* If it returns, x will contain the value in BC.
* If you want ROM code to run, observe normal precautions with IY and HL'.
  
Syntax is: bin2bas { -t=filename | -z=filename | -p } input { output }

  If no output filename is present, stdout will be used. The options are:

   -t=filename: Output the BASIC as a .TAP file containing the named program.
   -z=filename: Output the BASIC as a .ZXT file containing the named program.
   -p:          Output the BASIC in +3DOS format (default)

tapcat
------

  tapcat creates a .TAP file from one or more files (either +3DOS or raw). 
         It can also append to an existing .TAP file.

  Syntax is: tapcat { -N } { -H } { -a addr } <tapfile> <file> <file>

  If the -N option is present, then a new .TAP file is created. If not, the 
files will be appended to an existing .TAP file. 
  If the -H option is present, the files will be appended as headerless data
blocks.
  If the -a option is present, the load address specified will be used in
the tape header added to each file. Otherwise, a load address of 0 will be
used.

tapget
------

   tapget extracts one or more files from a .TAP file. 

Syntax is: tapget <tapfile> <file> <file> ...

   The filenames passed can include '*' and '?' as wildcards. Each specified
file will be saved in +3DOS format. Note that tapget can't extract headerless
blocks.

tapls
-----

  tapls lists the contents of one or more .TAP files.

Syntax is: tapls { -l | -3 } <tapfile> <tapfile> ...

  The options control how the output is formatted. If no options are present,
a simple list of names is displayed:

astroclone
loader    
pic       
[Headerless]

  -3 produces a listing similar to the +3DOS 'CAT "T:"' command:

Program: astroclone LINE 0 
Bytes: loader     CODE 24500,100
Bytes: pic        CODE 50000,7000
Headerless block: Type=255 length=40536

  -l produces a listing similar to the UNIX 'ls -l' command:

-r--r--r-- BASIC     190 astroclone
-r--r--r-- Bytes     100 loader    
-r--r--r-- Bytes    7000 pic       
-r--r--r-- [0xff]  40536 (Headerless block)

tapsplit
--------
 
  tapsplit explodes an entire .TAP file into its component parts, including
any headerless blocks. 

Syntax is: tapsplit <tapfile> { <output-dir> }

  If no output directory is supplied, the current directory will be used.

  Each file in the .TAP will be written to the chosen output directory. If its 
name collides with an existing file, ".1" will be appended and the process 
repeated.

sna2tap
-------

  sna2tap converts a Spectrum snapshot in the .SNA format to a .TAP file that
can be loaded on a 48k Spectrum.

Syntax is: sna2tap {options} { input-file { output-file } }

  Options are:

--base=addr: sna2tap overwrites 62-65 bytes of the snapshot with its 
            final-stage loader code. By default this is placed just below the
            bottom of the Z80 stack. Some games, particularly those which 
            use the stack for copying memory to and fro, may be corrupted by
            this; if so, you need to find somewhere else to put this code. In
            some cases the video RAM may be suitable; in others, you would 
            have to find a free space in the snapshot, perhaps using a hex 
            editor.
             The address can be specified in decimal or (if preceded by "0x") 
            hex. So --base=16384 or --base=0x4000 are equivalent.

--name=name: The name to give the BASIC loader (by default, "SNAPSHOT")
--tap:       Output in .TAP format (default)
--zxt:       Output in .ZXT format

  If the input and output filenames aren't specified, standard input and 
output are used. 

  To reverse the process, run tapsplit on the generated .TAP file. It will 
save the snapshot as "Block1.sna".

  (Technical note: tapls, tapsplit and tap2dsk assume that any block of type 
  0x53 ('S') and length 49179 bytes is a .SNA-format snapshot, and display it 
  accordingly.)

tap2dsk / mkp3fs
----------------

  tap2dsk and mkp3fs are both programs that generate a new +3DOS-format disc
image. mkp3fs creates a disc image and injects one or more files into it, 
while tap2dsk transfers the contents of a .TAP file into the new disc image.
  Don't pass the name of an existing disc image to either program, unless 
you want it overwritten without warning.

Syntax is: 
	mkp3fs  { options } dskfile filename filename ...
	tap2dsk { options } tapfile dskfile

  Both programs take the same options:

-180:               Generate a 180k disc image.
-720:               Generate a 720k disc image.
-cpmonly:	    Omit the FAT12 directory, FAT and boot sector (see below)
-dosonly:           Make the disc fully FAT12 (see below)
-nostamps:          Do not include file time stamps. You may want to use this
                    if you are writing headerless files to a 720k disc image,
                    because +3DOS appears to have difficulty reading 
		    headerless files in this situation.
-label <labelname>: The CP/M disc label. If none is specified, tap2dsk will 
		    use the name of the TAP file and mkp3fs won't create a 
		    label. If -nostamps is present, the label will not be 
                    written either.
-type <type>        The LibDsk type of the file to write (eg, dsk/edsk for a 
                    disc image, floppy for a real drive). If LibDsk wasn't
                    present at compile time, this option is ignored.
-compress <type>    The LibDsk compression method with which the new file 
                    should be compresssed. Again, if LibDsk wasn't around this
		    option is ignored.

Disc compatibility
~~~~~~~~~~~~~~~~~~
  By default, the disc images written by tap2dsk are readable not only as
CP/M discs under +3DOS, but also as FAT12 discs under Windows 9x, Linux and 
DOS. They are not, however, readable on Windows NT/2000/XP/Vista. Even under
those systems that can read them, there are a couple of caveats:

* If you've got Windows 95/98/ME, run the included DUALDOS.REG registry script
 to stop Windows' volume tracker scribbling on the boot sector.

* Changes made to the +3DOS directory won't update the PCDOS directory or 
 vice versa, so don't rely on this feature after writing to the disc.

Two options are supplied to control this feature:

-cpmonly  leaves out the DOS directory, FAT and boot sector. The disc will 
          still work as a CP/M disc under +3DOS, but not as a FAT12 disc under
          anything else.

-dosonly  writes a fully DOS-compatible boot sector. The disc will work 
          fine as a FAT12 disc under Windows NT/2000/XP/Vista, and won't 
          require the registry script to be run under Windows 95/98/ME, but
          it won't be usable on the +3.

  If both options are supplied, -cpmonly takes priority.

sna2dsk
-------

  sna2dsk takes a 48k .SNA-format snapshot and generates a +3DOS-format disc 
image that, run on a real Spectrum +3, ought to load that snapshot. Don't 
pass the name of an existing disc image to the program, unless you want it 
overwritten without warning.

Syntax is: 
	sna2dsk { options } snafile dskfile {custom-rom-file}

  If a custom ROM file is specified, the snapshot will be run in the 4-5-6-3
memory environment, with the specified ROM file in the bottom 16k.

-base addr:  sna2dsk overwrites 14 bytes of the snapshot with its
            final-stage loader code. By default this is placed just below the
            bottom of the Z80 stack. Some games, particularly those which 
            use the stack for copying memory to and fro, may be corrupted by
            this; if so, you need to find somewhere else to put this code. In
            some cases the video RAM may be suitable; in others, you would 
            have to find a free space in the snapshot, perhaps using a hex 
            editor.
             The address can be specified in decimal or (if preceded by "0x") 
            hex. So --base=16384 or --base=0x4000 are equivalent.

The following options only apply if a custom ROM file is _not_ in use:

-usr0:		    The resulting disc image will run the snapshot in USR0 
                    mode, with memory paging hardware enabled. This is the 
                    default. 

-48basic:	    The resulting disc image will run the snapshot in 48 BASIC
                    mode, with memory paging hardware disabled. 

The following options only apply if a custom ROM file _is_ in use:

-pause:             After loading all data but before launching the game, 
                    pause briefly. This was intended for Multiface 3 users,
                    so they could activate the Multiface before the Spectrum
                    switched to all-RAM mode, in which the Multiface can't
                    be used.

-noscreen:	    Leave the screen untouched when loading. This is intended
		    for users who want to edit the BASIC loader to add a
                    custom loading screen.

  The options from mkp3fs / tap2dsk are also supported:

-180:               Generate a 180k disc image.
-720:               Generate a 720k disc image.
-cpmonly:	    Omit the FAT12 directory, FAT and boot sector (see below)
-dosonly:           Make the disc fully FAT12 (see below)
-nostamps:          Do not include file time stamps. You will want to use this
                    if you are writing to a 720k disc image, because +3DOS 
                    appears to have difficulty reading headerless files (such
                    as .SNA snapshots) in this situation.
-label <labelname>: The CP/M disc label. 
-type <type>        The LibDsk type of the file to write (eg, dsk/edsk for a 
                    disc image, floppy for a real drive). If LibDsk wasn't
                    present at compile time, this option is ignored.
-compress <type>    The LibDsk compression method with which the new file 
                    should be compresssed. Again, if LibDsk wasn't around this
		    option is ignored.

dsk2tap
-------

  dsk2tap takes a .DSK image, and creates a .TAP file which, when run on a 
Spectrum +3, writes the contents of that image out to disc. It depends 
heavily on LibDsk and will not work without it.

Syntax is: 
      ./dsk2tap {options} dskfile tapfile

Options are:
-type <type>    The LibDsk type of the input disc image (default is to 
                autodetect).
-side <side>    Force side 0 or side 1 of input
-retry <count>  Set number of retries on error
-dstep          Double-step when reading
-stubborn       Ignore any read errors
-format         Force a specified LibDsk disc format (default is to 
                autodetect).

  The copy is simplistic; it only works on unprotected discs, and makes no 
attempt to replicate the original sector order.


ChangeLog
2011-03-27 John Elliott

	* sna2dsk: New program created: wraps a .SNA snapshot in an 
	 auto-booting +3 diskette image.

2010-05-03 John Elliott

	* taptools-1.0.7 released.

	* dsk2tap: Tweaked the BASIC loader so that if it runs to completion,
	 the program stops with 9 STOP rather than whatever error the +3DOS
	 loader came up with.

2010-04-17 John Elliott

	* taptools-1.0.6 released.

	* dsk2tap: Naively converts an unprotected .DSK to a .TAP file which,
	 when run, formats a floppy and creates a copy of the original disc.

2010-01-24 John Elliott

	* taptools-1.0.5 released.

	* tapcat: Added a -H option to append headerless blocks to the
	 tape file.

2008-02-09 John Elliott

	* taptools-1.0.4 released.

	* sna2tap: Naively converts a .SNA file to an autoloading .TAP file.

	* tapls, tapsplit, tap2dsk: Recognise tape blocks created by SNA2TAP. 
	Also fix an integer overflow issue on DOS systems with 16-bit integers.

2008-03-18 John Elliott

	* taptools-1.0.3 released.

	* bin2bas: New program created to wrap a BASIC loader round a 
	 binary blob.

2008-03-16 Garry Lancaster

	* specform, tapcat: -a option added to set load address.

	* tapget: Various bugfixes.

2007-06-28 John Elliott

	* taptools-1.0.2 released.

	* tap2dsk, mkp3fs: New utilities created to construct +3 disc images
	 from scratch.

2007-06-07 John Elliott

	* tapsplit: It seems that on or about 2005-07-11, one Javier Herrera
	 released a modified version of tapsplit but breached the GPL by not
	 distributing source (see
	 <1121094019.524192.7650@g43g2000cwa.googlegroups.com> ).
	 Created tapsplit-1.0.1, which should be a suitable substitute and 
	 can be distributed without fear of my armies of undead lawyers.

2005-06-30 John Elliott

	* taptools-1.0.0 released, comprising specform, unspecform,
	tapcat, tapget, tapls and tapsplit.

2000-06-10 John Elliott

	* tapsplit-1.0.0 released.

1996-12-14 John Elliott

	* Earliest known tapcat source.


mkp3fs.c
/************************************************************************

    TAPTOOLS v1.0.3 - Tapefile manipulation utilities

    Copyright (C) 1996, 2005, 2007  John Elliott <jce@seasip.demon.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "taputil.h"
#include "config.h"
#include "dskbits.h"
#include "plus3fs.h"

#ifdef __PACIFIC__
#define AV0 "MKP3FS"
#else
#define AV0 argv[0]
#endif


const char *dsktype = "dsk";
const char *compress = NULL;
const char *label = NULL;
int timestamp = 1;
int format = 180;
char fcbname[12];
PLUS3FS gl_fs;
int gl_uid = 0;
int cpmonly = 0;
int dosonly = 0;

typedef struct queued_file
{
	struct queued_file *next;
	const char *pathname;
	char filename[13];
} QFILE;

QFILE *qf_head = NULL;

void add_filename(const char *s)
{
	QFILE *qf = malloc(sizeof(QFILE));
	QFILE *qf2;
	char *sep;

	if (!qf)
	{
		fprintf(stderr, "Out of memory\n");
		exit(1);
	}
	qf->pathname = s;
	sep = strrchr(s, '/');
	if (sep) s = sep + 1;
#if defined(__MSDOS__) || defined(WIN32)
	sep = strrchr(s, '\\');
	if (sep) s = sep + 1;
	sep = strrchr(s, ':');
	if (sep) s = sep + 1;
#endif
	strncpy(qf->filename, s, 12);
	qf->filename[12] = 0;
	qf->next = NULL;
/* Add this entry to the end of the linked list */
	if (!qf_head)
	{
		qf_head = qf;
	}
	else
	{
		qf2 = qf_head;
		while (qf2->next != NULL) qf2 = qf2->next;
		qf2->next = qf;
	}
}

static unsigned char boot_180[] =
{
        0x00, /* Disc type */
        0x00, /* Disc geometry */
        0x28, /* Tracks */
        0x09, /* Sectors */
        0x02, /* Sector size */
        0x01, /* Reserved tracks */
        0x03, /* ?Sectors per block */
        0x02, /* ?Directory blocks */
        0x2A, /* Gap length (R/W) */
        0x52  /* Gap length (format) */
};

static unsigned char boot_720[] =
{
        0x03, /* Disc type */
        0x81, /* Disc geometry */
        0x50, /* Tracks */
        0x09, /* Sectors */
        0x02, /* Sector size */
        0x02, /* Reserved tracks */
        0x04, /* ?Sectors per block */
        0x04, /* ?Directory blocks */
        0x2A, /* Gap length (R/W) */
        0x52  /* Gap length (format) */
};


static void report(const char *s)
{
        printf("%s\r", s);
        fflush(stdout);
}

static void report_end(void) 
{
        printf("\r%-79.79s\r", "");
        fflush(stdout);
}

void diewith(dsk_err_t err)
{
	fprintf(stderr, "Cannot write %-8.8s.%-3.3s: %s\n",
			fcbname, fcbname + 8, p3fs_strerror(err));

	p3fs_umount(gl_fs);
	exit(1);
}


dsk_err_t put_file(QFILE *qf)
{
	FILE *fp;
	dsk_err_t err;
	PLUS3FILE fpo;
	int c;

	p3fs_83name(qf->filename, fcbname);
	fp = fopen(qf->pathname, "rb");
	if (!fp) return DSK_ERR_SYSERR;

	printf("Writing %-8.8s.%-3.3s\n", fcbname, fcbname+8);

	err = p3fs_creat(gl_fs, &fpo, gl_uid, fcbname);
	if (err) return err;

	while ((c = fgetc(fp)) != EOF)
	{
		err = p3fs_putc(c, fpo);
		if (err) return err;
	}
	fclose(fp);
	return p3fs_close(fpo);
}

void syntax(const char *av0)
{
	fprintf(stderr, "Syntax: %s {options} dskfile file file ...\n\n"
			"Options are:\n"
			"  -180: Write a 180k image (default)\n"
			"  -720: Write a 720k image\n"
			"  -cpmonly: Write a disc only usable by +3DOS\n"
			"  -dosonly: Write a disc only usable by PCDOS\n"
			"  -type type: Output file type (default=dsk)\n"
			"  -compress cmp: Output compression (default=none)\n"
			"  -label lbl: Set disk label (default=none)\n"
			"  -nostamps: Do not include file date stamps\n", av0);
      	 
	exit(1);
}

int main(int argc, char **argv)
{
	dsk_err_t err;
	int n, optend = 0;
	const char *dskfile = NULL;
	const char *option, *value;
	QFILE *qf;

	for (n = 1; n < argc; n++)
	{
		if (!strcmp(argv[n], "--"))
		{
			optend = 1;
			continue;
		}
		if (optend || argv[n][0] != '-')
		{
			if (dskfile == NULL) dskfile = argv[n];
			else add_filename(argv[n]);	
			continue;
		}
		option = argv[n] + 1;
		if (option[0] == '-') ++option;

		if      (!strcmp (option, "180")) 
		{
			format = 180;
			continue;
		}
		else if (!strcmp (option, "720"))
		{
			format = 720;
			continue;
		}
		else if (!strcmp (option, "nostamps")) 
		{
			timestamp = 0;
			continue;
		}
		else if (!strcmp (option, "cpmonly")) 
		{
			cpmonly = 1;	
			continue;
		}
		else if (!strcmp (option, "dosonly")) 
		{
			dosonly = 1;
			continue;
		}

/* Check for --variable=value and -variable value */
		value = strchr(option, '=');
		if (value) ++value;
		else
		{
			++n;
			if (n >= argc) syntax(AV0);
			value = argv[n];
		}
		if (!strncmp(option, "type", 4)) dsktype = value;
		else if (!strncmp(option, "label", 5)) label = value;
		else if (!strncmp(option, "compress", 8)) compress = value;
		else syntax(AV0);
	}
	if (dskfile == NULL || qf_head == NULL) 
	{
		syntax(AV0);
	}
	if (cpmonly && dosonly)
	{
		fprintf(stderr, "-cpmonly and -dosonly both present; -cpmonly "
				"takes priority\n");
	}

        dsk_reportfunc_set(report, report_end);

	err = p3fs_mkfs(&gl_fs, dskfile, dsktype, compress, 
			(format == 180) ? boot_180 : boot_720, timestamp);
	if (timestamp && !err && label)
	{
		p3fs_83name(label, fcbname);
		err = p3fs_setlabel(gl_fs, fcbname);
	}
	if (err)
	{
		fprintf(stderr, "Can't create %s: %s\n",
			dskfile, p3fs_strerror(err));
		exit(1);
	}
	for (qf = qf_head; qf != NULL; qf = qf->next)
	{
		err = put_file(qf);
		if (err) diewith(err);
	}
	if (cpmonly == 0 && (format == 180 || format == 720))
	{	
		err = p3fs_dossync(gl_fs, format, dosonly);
		if (err)
		{
			fprintf(stderr, "Cannot do DOS sync on %s: %s\n",
					dskfile, p3fs_strerror(err));
		
		}
	}
	err = p3fs_umount(gl_fs);
	if (err) 
	{
		fprintf(stderr, "Cannot close %s: %s\n",
				dskfile, p3fs_strerror(err));
		return 1;
	}
	return 0;
}


sna2tap.c
/************************************************************************

    TAPTOOLS v1.0.4 - Tapefile manipulation utilities

    Copyright (C) 1996, 2005, 2008, 2009 John Elliott <jce@seasip.demon.co.uk>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*************************************************************************/

#include "config.h"
#include "cnvshell.h"

typedef unsigned char byte;
typedef unsigned short word;

/* Length of final-stage loader that must be injected into the SNA image 
 * somewhere */
#define PRELUDE_MAX 65	/* If not immediately below the stack, needs 3 bytes
			   for a LD SP */
#define PRELUDE_MIN 62	/* If immediately below the stack, these can be
			   skipped */

static int st_prelude_size = PRELUDE_MAX;

/* Options */
static int st_base = -1;
static int st_zxt  = 0;
static char st_name[11] = "\020\007\021\000\023\001.sna";

/* The SNA, and the values of its registers */
static byte sna[49179];
static word i, r, hl1, de1, bc1, af1;
static word hl, de, bc, iy, ix;
static word af, sp, imm, iff, border;


char *cnv_set_option(int ddash, char *variable, char *value)
{
	if (variable[0] == 'n' || variable[0] == 'N')
	{
		if (!value) return "No filename supplied";
		sprintf(st_name, "%-10.10s", value);
		return NULL;
	}
	if (variable[0] == 'b' || variable[0] == 'B')
	{
		if (value && (isdigit(value[0]) || value[0] == '&'))
		{
			if ((value[0] == '0' && value[1] == 'x') ||
			    (value[0] == '0' && value[1] == 'X'))
			{
				if (!sscanf(value + 2, "%x", &st_base))
				{
					return "Invalid base address supplied";
				}
			}
			else if (value[0] == '&')
			{
				if (!sscanf(value + 1, "%x", &st_base))
				{
					return "Invalid base address supplied";
				}
			}
			else if (!sscanf(value, "%d", &st_base))
			{
				return "Invalid base address supplied";
			}
			return NULL;
		}
		return "No base address supplied";	
	}
	if (variable[0] == 't' || variable[0] == 'T')
	{
		st_zxt = 0;
		return NULL;
	}
	if (variable[0] == 'z' || variable[0] == 'Z')
	{
		st_zxt = 1;
		return NULL;
	}
	return "Unrecognised option";
}

static char helpbuf[2000];

char *cnv_progname = "sna2tap";

char *cnv_help(void)
{
	sprintf(helpbuf, "Syntax: %s { options } "
			"{ infile.SNA { outfile.TAP }}\n\n"
		"Options are:\n"
	/*	 1...5...10...15...20...25...30...35...40 */
		"--base: The address of an area of %d "
		"bytes which will be overwritten by the\n"
	        "        loader. If not specified, the ar"
		"ea just below the Z80 stack will be\n"
	        "        used.\n"
		"--name: The filename to use for the BASIC"
		" loader.\n"
		"--tap:  Output in .TAP format (default)\n"
		"--zxt:  Output in .ZXT format\n", 
		cnv_progname, st_prelude_size);

	return helpbuf;
}




word peek2(int a)
{
	word v = sna[a + 1];
	return (v << 8) | sna[a];
}

void poke2(byte *array, int a, word v)
{
	array[a] = (v & 0xFF);
	array[a+1] = (v >> 8);
}


void parse_sna(void)
{
	i   = sna[0]; i *= 0x101;
	hl1 = peek2(1);
	de1 = peek2(3);
	bc1 = peek2(5);
	af1 = peek2(7);
	hl  = peek2(9);
	de  = peek2(11);
	bc  = peek2(13);
	iy  = peek2(15);
	ix  = peek2(17);
	iff = (sna[19] & 1);
	r   = sna[20]; r *= 0x101;
	af  = peek2(21);
	sp  = peek2(23);
	imm = sna[25];
	border = sna[26]; border *= 0x101;
}


static byte launch[] =
{
/* Launch code, injected into memory. */
	0xE1, 0xD1, 0xC1, 0xF1,	/* +00 POP HL ! POP DE ! POP BC ! POP AF */
	0x08, 0xD9,		/* +04 EX AF, AF' ! EXX */
	0xFD, 0xE1, 0xDD, 0xE1, /* +06 POP IY ! POP IX */
	0xE1, 0xD1, 0xC1, 	/* +10 POP HL ! POP DE ! POP BC */
	0xF1, 0xD3, 0xFE,	/* +13 POP AF ! OUT (254),A */
	0xF1, 0xED, 0x47, 	/* +16 POP AF ! LD I,A */
	0xF1, 0xED, 0x4F,	/* +19 POP AF ! LD R,A */
	0xF1,			/* +22 POP AF */
	0x31, 0x00, 0x00,	/* +23 LD SP, nnnn */
	0xED, 0x00,		/* +26 IM 0/1/2 */
	0x00,			/* +28 EI or DI */
	0xED, 0x45		/* +29 RETN */	
				/* +31 */
};

/* The launch code must be followed by 34 bytes of stack:
 * 	DW 0, 0, 0		; +31 Workspace for tape loader
 * 	DW launch_code		; +37 
 * 	DW HL', DE', BC', AF'	; +39
 * 	DW IY, IX, HL, DE, BC	; +47
 * 	DW border, I, R, AF 	; +57 
 * 				; +65 */


static byte basic[] =
{
	0x00, 0x0A, 		/* +00 Line 10 */
	0x13, 0x00, 		/* +02 Length 19 */
	0xEA,			/* +04 REM */

	0xF3,			/* +05 DI */
	0x31, 0x00, 0x00,	/* +06 LD SP, nnnn */
	0xDD, 0x21, 0xE5, 0x3F,	/* LD IX, 3FE5h */
	0x11, 0x1B, 0xC0,	/* LD DE, 0C01Bh */
	0x3E, 'S',		/* LD A, 'S' */
	0x37,			/* SCF */
	0xC3, 0x56, 0x05,	/* JP 556h */	

	0x0D,				/* End of line 10 */
	0x00, 0x14, 			/* Line 20 */
	0x32, 0x00,			/* <len 50> */
	0xF1, 'x', '=', 0xC0, '(', '5',	/* LET x=USR (5 */
/*	0xF1, 'x', '=', 0x20, '(', '5',	 * LET x= (5         {for testing} */
	0x0E, 0, 0, 5, 0, 0, '+', 	/* [number 5] + */
	0xBE, '2', '3', '6', '3', '5',	/* PEEK 23635 */
	0x0E, 0, 0, 0x53, 0x5C, 0, '+',	/* [number 23635] + */
	'2', '5', '6',			/* + 256 */
	0x0E, 0, 0, 0, 1, 0, '*',	/* [number 256] * */
	0xBE, '2', '3', '6', '3', '6',	/* PEEK 23636 */
	0x0E, 0, 0, 0x54, 0x5C, 0,	/* [number 23636] */
	')', 0x0D	


};

static byte tapheader[17] = 
{
	0x00,		/* BASIC */
	'0', '1', '2', '3', '4', 
	'5', '6', '7', '8', '9', 
	0x00, 0x00,	/* program+vars size */
	0x00, 0x00,	/* autostart line */
	0x00, 0x00,	/* program size */
};

static byte zxtheader[128] =
{
	 'P',  'L',  'U',  'S',  '3',  'D',  'O', 'S',
	0x1A, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 'T',
	 'A',  'P',  'E',  'F',  'I',  'L',  'E'
};



char *mk_loader(int base)
{
	unsigned char prelude[PRELUDE_MAX];
	unsigned baseu;
	unsigned newsp;
	int n;

	memset(prelude, 0, sizeof prelude);
	if (base == -1) 
	{
		baseu = sp - PRELUDE_MIN;
		st_prelude_size = PRELUDE_MIN;
		fprintf(stderr, "Automatically placing loader code at 0x%x\n",
				baseu);
	}
	else
	{
		baseu = base;
		st_prelude_size = PRELUDE_MAX;
/* If they just so happened to pick the right place at the bottom of the
 * stack, take advantage of it */
		if (baseu == (sp - PRELUDE_MIN))
		{
			st_prelude_size = PRELUDE_MIN;
		}
/*		fprintf(stderr, "Manually placing loader code at 0x%x\n",
				baseu); */
	}
	if (baseu < 0x4000 || baseu >= (0xFFFF - st_prelude_size))
	{
		return "Loader code will not fit; specify a different "
			"address with --base.";
	}

/*	fprintf(stderr, "base=%04x\n", baseu); */

	memcpy(prelude, launch, sizeof(launch));

	poke2(prelude, 24, sp);
	switch (imm)
	{
		case 0: prelude[26] = 0xED; prelude[27] = 0x46; break;
		default:
		case 1: prelude[26] = 0xED; prelude[27] = 0x56; break;
		case 2: prelude[26] = 0xED; prelude[27] = 0x5E; break;
	}
	prelude[28] = (iff & 1) ? 0xFB : 0xF3;	/* EI or DI */

/* If we're at the bottom of the stack, don't need to load SP. Cut that
 * instruction out. */
	if (st_prelude_size == PRELUDE_MIN)
	{
		for (n = 23; n < PRELUDE_MIN; n++)
		{
			prelude[n] = prelude[n+3];
		}
		newsp = 34;
	}
	else newsp = 37;
	/* Point BASIC at our last-stage-launcher code */
	poke2(basic, 7, baseu + newsp + 2);
	poke2(prelude, newsp, baseu); newsp += 2;
	poke2(prelude, newsp, hl1); newsp += 2;
	poke2(prelude, newsp, de1); newsp += 2;
	poke2(prelude, newsp, bc1); newsp += 2;
	poke2(prelude, newsp, af1); newsp += 2;
	poke2(prelude, newsp, iy); newsp += 2;
	poke2(prelude, newsp, ix); newsp += 2;
	poke2(prelude, newsp, hl); newsp += 2;
	poke2(prelude, newsp, de); newsp += 2;
	poke2(prelude, newsp, bc); newsp += 2;
	poke2(prelude, newsp, border); newsp += 2;
	poke2(prelude, newsp, i); newsp += 2;
	poke2(prelude, newsp, r); newsp += 2;
	poke2(prelude, newsp, af); newsp += 2;


	memcpy(sna + baseu - 0x3fe5, prelude, st_prelude_size);
	return NULL;
}


char *save(FILE *fp, byte *addr, unsigned count)
{
	if ((unsigned)fwrite(addr, 1, count, fp) < count)
	{
		return "Cannot write to output file";
	}
	return NULL;
}

char *saveblock(FILE *fp, byte *data, unsigned count, byte type)
{
	byte blockhead[3];
	byte sum;
	unsigned n;
	char *boo;

	poke2(blockhead, 0, 2 + count);
	blockhead[2] = type;

	boo = save(fp, blockhead, 3); if (boo) return boo;
	boo = save(fp, data, count); if (boo) return boo;

	for (n = 0, sum = type; n < count; n++) sum ^= data[n];
	return save(fp, &sum, 1);
}


char *cnv_execute(FILE *infp, FILE *outfp)
{
	char *boo;

	if ((unsigned)fread(sna, 1, sizeof(sna), infp) < sizeof(sna))
	{
		return "Premature EOF on SNA file";
	}	
	parse_sna();
	boo = mk_loader(st_base); if (boo) return boo;

	memcpy(tapheader+1, st_name, 10);
	poke2(tapheader, 11, sizeof(basic));
	poke2(tapheader, 15, sizeof(basic));

	if (st_zxt)
	{
		int n;
		byte sum;

		poke2(zxtheader, 11, 128 + 
				     4 + sizeof(tapheader) + 
				     4 + sizeof(basic) +
				     4 + sizeof(sna));
		for (n = 0, sum = 0; n < 127; n++) sum += zxtheader[n];
		zxtheader[127] = sum;
		boo = save(outfp, zxtheader, 128);
		if (boo) return boo;

	}

/* Save the BASIC header */
	boo = saveblock(outfp, tapheader, sizeof(tapheader), 0);
	if (boo) return boo;

/* Save the BASIC block */
	boo = saveblock(outfp, basic, sizeof(basic), 0xFF);
	if (boo) return boo;

/* Save the SNA block */
	return saveblock(outfp, sna, sizeof(sna), 'S');
}

